llpc/lower/llpcSpirvLowerGlobal.cpp

```diff
@@ -263,86 +263,91 @@ bool SpirvLowerGlobal::runImpl(Module &module) {
+    if (function.getLinkage() == GlobalValue::InternalLinkage)
+      continue;
+    for (BasicBlock &block : function) {
+      Instruction *terminator = block.getTerminator();
+      if (!terminator || terminator->getOpcode() != Instruction::Ret)
+        continue;
+      ReturnInst *returnInst = dyn_cast<ReturnInst>(terminator);
+      assert(m_retBlock);
+      BranchInst::Create(m_retBlock, &block);
+      m_retInsts.insert(returnInst);
+    }
+  }
 }
 
 // =====================================================================================================================
-// Visits "call" instruction.
+// Handle "call" instructions.
 //
-// @param callInst : "Call" instruction
-void SpirvLowerGlobal::visitCallInst(CallInst &callInst) {
-  // Skip if "emit" and interpolation calls are not expected to be handled
-  if (!static_cast<bool>(m_instVisitFlags.checkEmitCall) && !static_cast<bool>(m_instVisitFlags.checkInterpCall))
-    return;
-
-  auto callee = callInst.getCalledFunction();
-  if (!callee)
-    return;
-
-  auto mangledName = callee->getName();
-
-  if (m_instVisitFlags.checkEmitCall) {
-    if (mangledName.startswith(gSPIRVName::EmitVertex) || mangledName.startswith(gSPIRVName::EmitStreamVertex))
-      m_emitCalls.insert(&callInst);
-  } else {
-    assert(m_instVisitFlags.checkInterpCall);
-
-    if (mangledName.startswith(gSPIRVName::InterpolateAtCentroid) ||
-        mangledName.startswith(gSPIRVName::InterpolateAtSample) ||
-        mangledName.startswith(gSPIRVName::InterpolateAtOffset) ||
-        mangledName.startswith(gSPIRVName::InterpolateAtVertexAMD)) {
-      // Translate interpolation functions to LLPC intrinsic calls
-      auto loadSrc = callInst.getArgOperand(0);
-      unsigned interpLoc = InterpLocUnknown;
-      Value *auxInterpValue = nullptr;
-
-      if (mangledName.startswith(gSPIRVName::InterpolateAtCentroid))
-        interpLoc = InterpLocCentroid;
-      else if (mangledName.startswith(gSPIRVName::InterpolateAtSample)) {
-        interpLoc = InterpLocSample;
-        auxInterpValue = callInst.getArgOperand(1); // Sample ID
-      } else if (mangledName.startswith(gSPIRVName::InterpolateAtOffset)) {
-        interpLoc = InterpLocCenter;
-        auxInterpValue = callInst.getArgOperand(1); // Offset from pixel center
+// @param checkEmitCall : Whether we should handle emit call or not
+// @param checkInterpCall : Whether we should handle interpolate call or not
+void SpirvLowerGlobal::handleCallInst(bool checkEmitCall, bool checkInterpCall) {
+  assert(checkEmitCall != checkInterpCall);
+
+  for (Function &function : m_module->functions()) {
+    auto mangledName = function.getName();
+    // We get all users before iterating because the iterator can be invalidated
+    // by interpolateInputElement
+    SmallVector<User *> users(function.users());
+    for (User *user : users) {
+      CallInst *callInst = dyn_cast<CallInst>(user);
+      if (callInst == nullptr)
+        continue;
```

Comment:
Can `callInst` ever be nullptr here? If not, you can change `dyn_cast` to `cast` and remove the `if`.

Reply:
nit: also, the coding style says to prefer `if (!pr)`